### DISCUSSION TITLE
Emit any TV remote event that targets focused `Pressable` & `Touchable` components

### DISCUSSION THREAD

**Discussion by mateusz1913** [react-native-tvos/react-native-tvos#679] [Feature Requests]
👋 I wanted to start discussion about introducing some API that will let act on TV remote events that target `Pressable` & `Touchable` components.

To handle TV remote events in TV apps built with React Native, we use `TVEventHandler` utility. It's also used internally to map TV remote `select` events to `onPress` callbacks inside of `Pressable` & `Touchable` components.

Looking at the internal JS logic, each [`Pressable`](https://github.com/react-native-tvos/react-native-tvos/blob/43e88d57611a42a60578201f041b710c1e77376e/packages/react-native/Libraries/Components/Pressable/Pressable.js#L378) or [`Touchable`](https://github.com/react-native-tvos/react-native-tvos/blob/43e88d57611a42a60578201f041b710c1e77376e/packages/react-native/Libraries/Components/Touchable/TVTouchable.js#L39) handles `focus`/`blur`/`select`/`longSelect`.

The idea would be to use that internal logic to expose any TV remote event that `TVEventHandler` will emit.
For tvOS it would mean `playPause` && `longPlayPause` events and for AndroidTV there're a lot of events e.g. `rewind`, `fastForward`, `play`, `pause`, `playPause`, etc.

From my investigation, it seems that for AndroidTV platforms, only JS changes are needed. For tvOS to make JS changes work, additional ObjC gesture recognizers (like the one for [`select` button](https://github.com/react-native-tvos/react-native-tvos/blob/43e88d57611a42a60578201f041b710c1e77376e/packages/react-native/React/Views/RCTTVView.m#L79)) would have to be registered to handle `playPause` events on `RCTTVView` and `RCTViewComponentView` classes.

The open question would be how to expose new events:
- 1st approach would be to expose everything via existing `onPress`/`onLongPress` APIs
  - in theory that should not be a breaking change, because these callbacks already take whole `select` (or `longSelect`) event as an argument, however existing usages may rely only on `select` events without even checking the argument passed to callback
- 2nd approach would be to introduce new prop called `onTVRemotePress` that would handle every "press" event apart from `select`/`longSelect`

I'd personally prefer the 2nd approach as a safer one, but would love to discuss it!

**Comment by yusufyildirim** [comment]
Thanks for opening up a discussion!

> The idea would be to use that internal logic to expose any TV remote event that TVEventHandler will emit.
For tvOS it would mean playPause && longPlayPause events and for AndroidTV there're a lot of events e.g. rewind, fastForward, play, pause, playPause, etc.

I didn't get this part entirely. Are rewind, fastForward, play etc. considered as Remote events? I'm not sure about tvOS but I don't think they are on Android. There are separate remote keys for each of them. There's also some other stuff that maps left/right keys to fastForward/rewind actions but those are not considered as Remote events in that case as far as I know. Those are contextual and only make sense if there's a player in the app so I feel like it's Player's responsibility, not the remote handler's. Could you elaborate on this one a bit more? Maybe I'm missing something.

An extra thing to add, I'll very likely be changing the implementation of focus and press/select handling really soon and remove the TVRemoteHandler related logic for these purposes entirely to support this: https://github.com/react-native-tvos/react-native-tvos/discussions/668

**Comment by mateusz1913** [comment]
Yes, in the `ReactAndroidHWInputDeviceHelper` [util](https://github.com/react-native-tvos/react-native-tvos/blob/43e88d57611a42a60578201f041b710c1e77376e/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/modules/core/ReactAndroidHWInputDeviceHelper.java#L41) there're already multiple [`KeyEvent`](https://developer.android.com/reference/android/view/KeyEvent) key codes registered as a `KEY_EVENTS_ACTIONS` & `KEY_EVENTS_LONG_PRESS_ACTIONS` maps. So definitely no need to add any more `KeyEvent` there, just make it possible to listen to those events without writing additional `TVEventHandler` instances in JS codebase.

> Are rewind, fastForward, play etc. considered as Remote events? I'm not sure about tvOS but I don't think they are on Android. There are separate remote keys for each of them.

By remote event, I meant events emitted by clicking buttons that are on remote, if I used wrong terminology, then apologies for the confusion 😅 . tvOS offers play/pause button on its remotes, while in AndroidTV world, there are FireTV remotes and Nvidia Shield remotes which offer rewind, fast-forward and play/pause buttons

About mapping left/right buttons to rewind/fast-forward, I agree with what you mentioned, but I'd say it's out-of-scope for that proposal.
